### DISCUSSION TITLE
Partial re-renders

### DISCUSSION THREAD

**Discussion by ShaMan123** [fabricjs/fabric.js#7724] [Ideas]
@asturur @melchiar 

The new group is maturing nicely. Thanks to that it is now possible to explore a few interesting directions.
Partial re-rendering. If we split up the canvas to 3X3 (or nXn) groups with `objectCaching` set to true we basically got it.
Related https://github.com/fabricjs/fabric.js/discussions/7506#discussioncomment-2218293

Layering is another, https://github.com/fabricjs/fabric.js/issues/7692#issuecomment-1045909212

Object batching is another but it is quite similar to partial re-rendering. A different name and a different need, but with group it's the same implementation.
Take a look at #7675

**Comment by asturur** [comment]
yes i want to work on partial re-renders since long, i hope i get to do it at some point.
I have an idea in mind i have to understand how performant is when you have to calculate object position/intersections in order to detect which objects do actually need to be redrawn.

The issue with the example of the 10.000 objects is that looping over 10.000 things to perform various checks, is already a show stopper for 60fps drawing ( of course redrawing just a couple of them would help )

**Reply by Ch1sKey** [reply]
The problem with collision intersections check is well known. The common solution is to use spacial data-structures, which speeds up search of the objects by coordinates. I see its quite challenging to embed such a structure to the current architecture. However potential performance gain should worth it.

**Comment by Ch1sKey** [comment]
As pointed here #8581, in theory, "partial re-renders" aka "advanced caching" aka "drawing by tiles" might be the only and most efficient way to scale fabric performance. The slowest part of rendering now is a render of each individual object, which would scale proportionally O(n) no matter how hard we try to optimise it.

**Reply by ShaMan123** [reply]
I disagree.
A simple example disproving it:
dragging an object across the canvas really quick.
You will need to invalidate all tiles...
It is a complex topic and I am not sure there is a single solution for all cases.
We have discussed exposing an interface for devs to do that themselves.

**Reply by Ch1sKey** [reply]
It's not necessary to invalidate all tiles while dragging but I see your point. There are multiple solutions for such problems with their own drawbacks.
